getgather.py

```diff
@@ -0,0 +1,322 @@
+
+import nanoid
+import pwinput
+from bs4 import BeautifulSoup
+from bs4.element import Tag
+from patchright.async_api import BrowserContext, Page, async_playwright
+
+from getgather.distill import Pattern, distill
+
+NORMAL = "\033[0m"
+BOLD = "\033[1m"
+YELLOW = "\033[93m"
+MAGENTA = "\033[35m"
+RED = "\033[91m"
+GREEN = "\033[92m"
+CYAN = "\033[36m"
+
+ARROW = "⇢"
+CROSS = "✘"
+
+FRIENDLY_CHARS = "23456789abcdefghijkmnpqrstuvwxyz"
+
+
+async def sleep(seconds: float):
+    await asyncio.sleep(seconds)
+
+
+async def ask(message: str, mask: str | None = None) -> str:
+    if mask:
+        return pwinput.pwinput(f"{message}: ", mask=mask)
+    else:
+        return input(f"{message}: ")
+
+
+async def click(page: Page, selector: str, timeout: int = 3000) -> None:
+    LOCATOR_ALL_TIMEOUT = 100
+    locator = page.locator(selector)
+    try:
+        elements = await locator.all()
+        print(f'Found {len(elements)} elements for selector "{selector}"')
+        for element in elements:
+            print("Checking", element)
+            if await element.is_visible():
+                print("Clicking on", element)
+                try:
+                    await element.click()
+                    return
+                except Exception as err:
+                    print(f"Failed to click on {selector} {element}: {err}")
+    except Exception as e:
+        if timeout > 0 and "TimeoutError" in str(type(e)):
+            print(f"retrying click {selector} {timeout}")
+            await click(page, selector, timeout - LOCATOR_ALL_TIMEOUT)
+            return
+        raise e
+
+
+def search(directory: str) -> list[str]:
+    results: list[str] = []
+    for root, _, files in os.walk(directory):
+        for file in files:
+            results.append(os.path.join(root, file))
+    return results
+
+
+def parse(html: str):
+    return BeautifulSoup(html, "html.parser")
+
+
+class Handle(TypedDict):
+    id: str
+    hostname: str
+    location: str
+    context: BrowserContext
+    page: Page
+
+
+async def init(location: str = "", hostname: str = "") -> Handle:
+    global playwright_instance, browser_instance
+
+    id = nanoid.generate(FRIENDLY_CHARS, 6)
+    directory = f"data/profiles/{id}"
+
+    if not playwright_instance:
+        playwright_instance = await async_playwright().start()
+        browser_instance = await playwright_instance.chromium.launch(
+            headless=False, channel="chromium"
+        )
+
+    context = await playwright_instance.chromium.launch_persistent_context(  # type: ignore
+        directory, headless=False, viewport={"width": 1920, "height": 1080}
+    )
+
+    page = context.pages[0] if context.pages else await context.new_page()
+    return {"id": id, "hostname": hostname, "location": location, "context": context, "page": page}
+
+
+def load_patterns() -> list[Pattern]:
+    patterns: list[Pattern] = []
+    for name in [f for f in search("./getgather/connectors/brand_specs") if f.endswith(".html")]:
```

Comment:
can we use [glob](https://docs.python.org/3/library/glob.html) instead custom search function? 

```py
def load_patterns() -> list[Pattern]:
    patterns: list[Pattern] = []
    for name in glob.glob("./getgather/connectors/brand_specs/**/*.html", recursive=True):
        with open(name, "r", encoding="utf-8") as f:
            content = f.read()
        patterns.append({"name": name, "pattern": parse(content)})
    return patterns
```

Reply:
Great point, thanks @broerjuang!
I'll do that as a follow up, and for Middleman as well.